### PR TITLE
[IEnum] make_range<TStoredType> 

### PR DESCRIPTION
**Pull request by asklar**
Don't allow raw output type access; either it's a COM pointer, or you need to pass the type you want to hold (like unique_idlist)
see https://github.com/microsoft/wil/pull/398#discussion_r1414631422

### CODE REVIEW

**Review by ChrisGuzak — APPROVED** [review]

File: include/wil/com.h [hunk 1]
```diff
@@ -3122,27 +3123,34 @@ struct com_iterator
     }
 };
 
-
-template<typename IEnumXxx, wistd::enable_if_t<wil::details::has_next_v<IEnumXxx*>, int> = 0>
+template<typename TStoredType = void, typename IEnumXxx = void, wistd::enable_if_t<wil::details::has_next_v<IEnumXxx*>, int> = 0>
```
[the remark this answers is not in the document]
Comment (asklar) [reply include/wil/com.h:3126]: fixed - I had thought default template args behaved like function params where once you specified a default value, all subsequent values need defaults

File: include/wil/com.h [hunk 2]
```diff
@@ -3044,9 +3044,9 @@ namespace details
     struct com_enumerator_traits
     {
         using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
-        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, we use the raw pointer type IFoo*.
+        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, smart_result is void.
         using smart_result = wistd::conditional_t<wistd::is_pointer_v<Result> && wistd::is_base_of_v<::IUnknown, wistd::remove_pointer_t<Result>>,
-            wil::com_ptr<wistd::remove_pointer_t<Result>>, Result>;
+            wil::com_ptr<wistd::remove_pointer_t<Result>>, void>;
```
[the remark this answers is not in the document]
Comment (asklar) [reply include/wil/com.h:3049]: I tried doing this but it makes some of the tests impossible to write (I think), as I can't seem to sfinae to validate that make_range of something like IEnumIDList (without a smart holder type) results in an undefined type :-/

**Comment by asklar** [comment]
happy new year! @jonwis / @dunhor any concerns? if not can you merge? Thanks!

**Review by jonwis — APPROVED** [review]
Looks OK. It'd be nice if the deduction system could know about scalars, or had an escape hatch for structures of scalars. No change required now, just something for future noodling.

**Pull request merged**